ts/src/encoding/fastPforEncoder.ts

```diff
@@ -0,0 +1,401 @@
+    return next;
+}
+
+function ensureUint8Capacity(buffer: Uint8Buf, requiredLength: number): Uint8Buf {
+    if (requiredLength <= buffer.length) return buffer;
+
+    let newLength = buffer.length === 0 ? 1 : buffer.length;
+    while (newLength < requiredLength) {
+        newLength *= 2;
+    }
+
+    const next = new Uint8Array(newLength) as Uint8Buf;
+    next.set(buffer);
+    return next;
+}
+
+/**
+ * Internal workspace for the FastPFOR encoder.
+ * Not exported - workspace API is kept simple for internal use.
+ */
+type FastPforEncoderWorkspace = {
+    dataToBePacked: Int32Array[];
+    dataPointers: Int32Array;
+    byteContainer: Uint8Buf;
+    freqs: Int32Array;
+    best: Int32Array;
+};
+
+/**
+ * Generic bit-packing of 32 integers, matching JavaFastPFOR BitPacking.fastpack ordering.
+ * Writes exactly `bitWidth` int32 words into `out` starting at `outPos`.
+ */
+function fastPack32(inValues: Int32Array, inPos: number, out: Int32Buf, outPos: number, bitWidth: number): void {
+    if (bitWidth === 0) return;
+    if (bitWidth === 32) {
+        out.set(inValues.subarray(inPos, inPos + 32), outPos);
+        return;
+    }
+
+    const mask = MASKS[bitWidth] >>> 0;
+    let outputWordIndex = outPos;
+    let bitOffset = 0;
+    let currentWord = 0;
+
+    for (let i = 0; i < 32; i++) {
+        const value = (inValues[inPos + i] >>> 0) & mask;
+
+        if (bitOffset + bitWidth <= 32) {
+            currentWord |= value << bitOffset;
+            bitOffset += bitWidth;
+
+            if (bitOffset === 32) {
+                out[outputWordIndex++] = currentWord | 0;
+                bitOffset = 0;
+                currentWord = 0;
+            }
+        } else {
+            const lowBits = 32 - bitOffset;
+            const lowMask = MASKS[lowBits] >>> 0;
+            currentWord |= (value & lowMask) << bitOffset;
+            out[outputWordIndex++] = currentWord | 0;
+            currentWord = value >>> lowBits;
+            bitOffset = bitWidth - lowBits;
+        }
+    }
+
+    if (bitOffset !== 0) {
+        out[outputWordIndex] = currentWord | 0;
+    }
+}
+
+const pSize = normalizePageSize(DEFAULT_PAGE_SIZE);
+
+const initialPackedSize = (pSize / 32) * 4;
+const byteContainerSize = ((3 * pSize) / BLOCK_SIZE + pSize) | 0;
+
+function createEncoderWorkspace(): FastPforEncoderWorkspace {
+    const dataToBePacked: Int32Array[] = new Array(33);
+    for (let k = 1; k < dataToBePacked.length; k++) {
+        dataToBePacked[k] = new Int32Array(initialPackedSize);
+    }
+
+    return {
+        dataToBePacked,
+        dataPointers: new Int32Array(33),
+        byteContainer: new Uint8Array(byteContainerSize) as Uint8Buf,
+        freqs: new Int32Array(33),
+        best: new Int32Array(3),
+    };
+}
+
+/**
+ * Default encoder workspace, allocated lazily on first use.
+ * The encoder is used primarily for tests and is not safe for concurrent calls.
+ */
+let defaultEncoderWorkspace: FastPforEncoderWorkspace | undefined;
```

Comment:
Is this a singleton variable? What will happen if someone tries to use fastPFor in parallel?
I don't understand this variable definition to be honest...